src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -152,6 +169,33 @@ public void run() {
         }
     }
 
+    @Override
+    public synchronized void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+        // We block the cleaner thread to prevent race conditions between this
+        // thread and checkpointing thread invoking clean().
+        // When the cleanup starts in cleaner thread the checkpoint will skip
+        // it, but without waiting for the cleanup to finish (which might be
+        // critical for the checkpoint, e.g. closing FDs).
+        // The limitation is that code performing C/R must not wait on any task
+        // completed by the cleaner.
```

Comment:
Could you elaborate why there is this limitation?

Reply:
The thread is blocked until `afterRestore`. PhantomCleanableRefs are cleaned independently within its own priority class (after the thread is stopped), but any other cleanup tasks simply won't happen. Therefore if C/R after this point waits for the cleanup, it will deadlock.